pkg/memory-monitor/src/monitor/cgroups.c

```diff
@@ -0,0 +1,332 @@
+// Copyright (c) 2024 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/select.h>
+#include <syslog.h>
+#include <unistd.h>
+
+#include "config.h"
+#include "event.h"
+#include "monitor.h"
+#include "util.h"
+
+#include "cgroups.h"
+
+unsigned long cgroup_get_memory_usage(const char *cgroup_name) {
+    char path[PATH_MAX + 1];
+    int fd;
+    char str_usage[256];
+    unsigned long usage_bytes;
+
+    snprintf(path, sizeof(path), "%s/%s/memory.usage_in_bytes", CGROUP_PATH_PREFIX, cgroup_name);
+    fd = open(path, O_RDONLY);
+    if (fd == -1) {
+        syslog(LOG_ERR, "opening usage_in_bytes: %s", strerror(errno));
+        return 0;
+    }
+
+    ssize_t nread = read(fd, str_usage, sizeof(str_usage) - 1);
+    if (nread <= 0) {
+        syslog(LOG_ERR, "reading usage_in_bytes: %s", strerror(errno));
+        close(fd);
+        return 0;
+    }
+
+    str_usage[nread] = '\0';
+    bool error;
+    usage_bytes = strtoudec(str_usage, &error);
+    if (error) {
+        syslog(LOG_ERR, "strtoul: %s", strerror(errno));
+        close(fd);
+        return 0;
+    }
+
+    close(fd);
+    return usage_bytes;
+}
+
+
+unsigned long cgroup_get_total_cache(const char *cgroup_name) {
+    char buf[256];
+    FILE *file;
+    char line[256];
+    unsigned long total_cache = 0;
+
+    // Construct the path to the memory.stat file
+    snprintf(buf, sizeof(buf), "%s/%s/memory.stat", CGROUP_PATH_PREFIX, cgroup_name);
+
+    // Open the file
+    file = fopen(buf, "r");
+    if (file == NULL) {
+        syslog(LOG_ERR, "opening memory.stat: %s", strerror(errno));
+        return 0;
+    }
+
+    // Read the file line by line
```

Comment:
Are we sure that lines in memory.stat file won't be longer than 256 bytes? Otherwise we will read line in multiple chunks, or can we guarantee that it'll be like "\n\rtotal_cache". And also that the number after will be 256 - strlen("total cache ")? Because we strtoudec will consider NULL as an error

Reply:
Yes, we can guarantee. The format of the file is standard and we know what to expect there.
And about the size:
256 - 12 = 244. Even 240... 10**240 Kb... That's astronomical amount of memory. Yes, we will fit.